test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java

```diff
@@ -52,4 +59,161 @@ public void writeTo(StreamOutput out) throws IOException {
             throw new IllegalArgumentException("Not supported.");
         }
     }
+
+    public void testAssertXContentEquivalent() throws IOException {
+        XContentBuilder original = JsonXContent.contentBuilder();
+        original.startObject();
+        addRandomNumberOfFields(original);
+        {
+            original.startObject(randomAlphaOfLength(10));
+            addRandomNumberOfFields(original);
+            original.endObject();
+        }
+        {
+            original.startArray(randomAlphaOfLength(10));
+            int elements = randomInt(5);
+            for (int i = 0; i < elements; i++) {
+                original.value(randomAlphaOfLength(10));
+            }
+            original.endArray();
+        }
+        original.endObject();
+
+        XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY, original.bytes(), original.contentType());
+        XContentBuilder copy = JsonXContent.contentBuilder();
+        parser.nextToken();
+        XContentHelper.copyCurrentStructure(copy.generator(), parser);
+        assertToXContentEquivalent(original.bytes(), copy.bytes(), original.contentType());
+    }
+
+    public void testAssertXContentEquivalentErrors() throws IOException {
+        {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            {
+                builder.startObject("foo");
+                {
+                    builder.field("field1", "value1");
+                    builder.field("field2", "value2");
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+
+            XContentBuilder otherBuilder = JsonXContent.contentBuilder();
+            otherBuilder.startObject();
+            {
+                otherBuilder.startObject("foo");
+                {
+                    otherBuilder.field("field1", "value1");
+                }
+                otherBuilder.endObject();
+            }
+            otherBuilder.endObject();
+            AssertionError error = expectThrows(AssertionError.class,
+                    () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
+            assertEquals("different field names at [foo]: [field1, field2] vs. [field1] expected:<2> but was:<1>",
```

Comment:
I'd like to not only have the different map sizes in the error message but also which keys of the map (which xContent fields) are either missing or to much in the actual and the expected map, and the xContent path that leads to the map that is different. That way when debugging a test faillure we can quickly see e.g. which fields we didn't parse correctly etc....
Maybe rephrasing to: `different xContent at path [foo]. Expected [field1, field2], but was [field1]. Number of fields expected:<2> but was:<1>` would help?